chemprop/cli/train.py

```diff
@@ -542,170 +531,220 @@ def main(args):
+                else train_dset.Y.shape[1],
+                hidden_dim=args.ffn_hidden_dim,
+                n_layers=args.ffn_num_layers,
+                dropout=args.dropout,
+                activation=args.activation,
+                criterion=criterion,
+                n_classes=args.multiclass_num_classes,
+                spectral_activation=args.spectral_activation,
+            )
+
+            if isinstance(predictor_ffn, RegressionFFN):
+                scaler = train_dset.normalize_targets()
+                val_dset.normalize_targets(scaler)
+                logger.info(f"Train data: loc = {scaler.mean_}, scale = {scaler.scale_}")
+            else:
+                scaler = None
+
+            train_loader = data.MolGraphDataLoader(train_dset, args.batch_size, args.num_workers)
+            val_loader = data.MolGraphDataLoader(
+                val_dset, args.batch_size, args.num_workers, shuffle=False
+            )
+
+            if multicomponent:
+                if len(test_data[0]) > 0:
+                    test_dsets = [make_dataset(data, args.rxn_mode) for data in test_data[fold_idx]]
+                    test_dset = data.MulticomponentDataset(test_dsets)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
+            else:
+                if len(test_data) > 0:
+                    test_dset = make_dataset(test_data[fold_idx], args.rxn_mode)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
+
+            mpnn_cls = MulticomponentMPNN if multicomponent else MPNN
+            model = mpnn_cls(
+                mp_block,
+                agg,
+                predictor_ffn,
+                True,
+                None,
+                args.task_weights,
+                args.warmup_epochs,
+                args.init_lr,
+                args.max_lr,
+                args.final_lr,
+            )
+            logger.info(model)
+
+            monitor_mode = "min" if model.metrics[0].minimize else "max"
+            logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+
+            if no_cv:
+                output_dir = args.output_dir
+            else:
+                output_dir = args.output_dir / f"fold_{fold_idx}"
+
+            tb_logger = TensorBoardLogger(output_dir / "tb_logs")
+            checkpointing = ModelCheckpoint(
+                output_dir / "chkpts",
+                "{epoch}-{val_loss:.2f}",
+                "val_loss",
+                mode=monitor_mode,
+                save_last=True,
+            )
+
+            early_stopping = EarlyStopping("val_loss", patience=5, mode=monitor_mode)
```

Comment:
RE: our discussion during the hackathon last week. v1 would never actually _stop_ training, it would just load the best model from all epochs, acting as if it had stopped there. This will be different, in that v2 will actually _stop training_. Should we make this _very_ clear somewhere? Is this intended?

Reply:
We should make this configurable. I think I just added this in when I originally scaffolded out the training CLI from other code I've written. Long term, it would be nice if we could have a separate config file for callbacks, but for now, we can just add a `--patience` arg and set the default to `None` to indicate training for the full number of epochs:
```python
patience = args.patience if args.patience is not None else args.epochs
```

Reply:
Thanks for the suggestions! I have added a `--patience` arg and set the default to `None` as suggested.